GAGS coloring for Atmos that displays changes in temperature

timothymtorres opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

This PR adds temperature based coloring for atmos gas meters and freezers.

![meter_temps](https://user-images.githubusercontent.com/5195984/124202799-ccc4bc80-daa0-11eb-9be5-643684a069f4.png)

- [x] GAGS for gas meters
- [x] GAGS for freezers

What this does is change the color based on temperature.  There are 8 color states.

-  Grey (No pressure/No Temp/Unpowered/Broken)
- +700K Red
- +460K Orange
- +340K Yellow
- +270K Green (room temp)
- +200K Cyan
- +120K Blue
- +++1K Purple

These values are based on the BODYTEMP icon warnings that you get when a room is too hot or too cold.  There is also some minor code improvements where some of the icon heat and cold warnings were hardcoded values.  I made this a constant in the `__DEFINES.atmospherics.dm` file that you can see below.  These values have been applied to air alarms, thermomachines, gas meters, and species code.

```
// Body temperature warning icons
/// The temperature the red icon is displayed.
#define BODYTEMP_HEAT_WARNING_3 (BODYTEMP_HEAT_DAMAGE_LIMIT + 360) //+700k
/// The temperature the orange icon is displayed.
#define BODYTEMP_HEAT_WARNING_2 (BODYTEMP_HEAT_DAMAGE_LIMIT + 120) //460K
/// The temperature the yellow icon is displayed.
#define BODYTEMP_HEAT_WARNING_1 (BODYTEMP_HEAT_DAMAGE_LIMIT) //340K
/// The temperature the light green icon is displayed.
#define BODYTEMP_COLD_WARNING_1 (BODYTEMP_COLD_DAMAGE_LIMIT) //270k
/// The temperature the cyan icon is displayed.
#define BODYTEMP_COLD_WARNING_2 (BODYTEMP_COLD_DAMAGE_LIMIT - 70) //200k
/// The temperature the blue icon is displayed.
#define BODYTEMP_COLD_WARNING_3 (BODYTEMP_COLD_DAMAGE_LIMIT - 150) //120k
```

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

This is a QoL update for atmos tech nerds.  Instead of having to manually examine a gas meter or use an analyzer to determine what temperature is in the pipes you can rely on the colored icons to communicate that information rapidly.  This should also help prevent harmful gas leaks where a person unwrenches a pipe without realizing there is hot gas inside it.  And you should be able to troubleshoot a delaming SM faster and easier with these colors.  


## Changelog
:cl:
qol: Gas meters, freezers, and air alarms all now use GAGS coloring to display colors based on temperature. (red is hot, green is room temp, and purple is cold)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

timothymtorres committed:
Add greyscale to atmos meters

timothymtorres committed:
Add some changes to meter

timothymtorres committed:
Add constants for heat and cold warning icons

timothymtorres committed:
Add GAGS to atmos gas meters for temperature

timothymtorres committed:
Fix meters with no power or pressure to revert to gray

timothymtorres committed:
Change the flashing button texture for button4

timothymtorres committed:
Typo change color comments for BODYTEMP constants

LemonInTheDark commented:
![image](https://user-images.githubusercontent.com/58055496/124062682-f967ce00-d9e5-11eb-82f1-f8783c910311.png)
Gonna have to do something about this I think, I wonder why it's a limit in the first place? @ninjanomnom thoughts?

timothymtorres commented:
@LemonInTheDark I think #59923 is the solution.

LemonInTheDark commented:
Ah I missed that, sorry for the trouble

timothymtorres committed:
Fix meters icons to be compatible with mapping

timothymtorres committed:
Add GAGS to thermomachine for temperature

timothymtorres committed:
Change GAGS temperature color light green to be replaced with purple

timothymtorres committed:
Typo purple color for GAGS missing a #

timothymtorres commented:
Updates:

- I had to drop air alarm support since a spriter told me he was remaking all the wall mounted sprites.  May come back to this later.
- Decided to remove the light green color and add purple instead.  So from room temp (green) to coldest it is cyan, blue, and purple in that order.

@LemonInTheDark I'm waiting for @ninjanomnom GAGS PRs to get merged and then once those are done I think I need to update the mapping icons and then this will be ready to go!

timothymtorres committed:
Fix icon states to use JSON GAGS that are generated

ninjanomnom commented:
> @LemonInTheDark 
> 
> Gonna have to do something about this I think, I wonder why it's a limit in the first place? @ninjanomnom thoughts?

The fix as previously mentioned is in a pr already but as for why it exists: The limitation is basically to act as a reasonable limit to the size of individual config files. If you're going bigger than 50 layers for a single icon state you very likely have steps in there that could be reused. The limit doesn't actually prevent you from making more complex icons you just need to break up the config a bit and use reference layers.

timothymtorres committed:
Merge branch 'master' into atmos_mapping_upgrades

timothymtorres committed:
Fix GAGS atmos icons to be mapper friendly

timothymtorres commented:
@LemonInTheDark This is ready to go.  If @ninjanomnom finishes his autogenerate PR then he can ping me and I'll combine it with the atmos icons.

LemonInTheDark commented on code/modules/atmospherics/machinery/airalarm.dm:
```diff
@@ -85,7 +85,7 @@
 	var/list/TLV = list( // Breathable air.
 		"pressure" = new/datum/tlv(HAZARD_LOW_PRESSURE, WARNING_LOW_PRESSURE, WARNING_HIGH_PRESSURE, HAZARD_HIGH_PRESSURE), // kPa. Values are min2, min1, max1, max2
-		"temperature" = new/datum/tlv(T0C, T0C+10, T0C+40, T0C+66),
+		"temperature" = new/datum/tlv(BODYTEMP_COLD_WARNING_1, BODYTEMP_COLD_WARNING_1+10, BODYTEMP_HEAT_WARNING_1-27, BODYTEMP_HEAT_WARNING_1),
```
Undocced change reeeeeeeee

timothymtorres replied:
I said removed the hardcoded values for the BODYTEMP icon warnings in several places.  I will specifically mention it in the PR but it's also apart of ad502a97d6bd7e4585aaa40342b3b2351e8c8288

timothymtorres replied:
This is the edited version of the PR:

> These values are based on the BODYTEMP icon warnings that you get when a room is too hot or too cold. There is also some minor code improvements where some of the icon heat and cold warnings were hardcoded values. I made this a constant in the __DEFINES.atmospherics.dm file that you can see below. **These values have been applied to air alarms, thermomachines, gas meters, and species code.**

The bolded section is the new mention of the air alarms included.

LemonInTheDark commented on code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm:
```diff
@@ -81,14 +92,21 @@
+	if(target_temperature <= BODYTEMP_COLD_WARNING_3)
+		set_greyscale(colors=list(thermo_colors["purple"]))
+	else if(target_temperature <= BODYTEMP_COLD_WARNING_2)
+		set_greyscale(colors=list(thermo_colors["blue"]))
+	else if(target_temperature <= BODYTEMP_COLD_WARNING_1)
+		set_greyscale(colors=list(thermo_colors["cyan"]))
+	else if(target_temperature <= BODYTEMP_HEAT_WARNING_1)
+		set_greyscale(colors=list(thermo_colors["green"]))
+	else if(target_temperature <= BODYTEMP_HEAT_WARNING_2)
+		set_greyscale(colors=list(thermo_colors["yellow"]))
+	else if(target_temperature <= BODYTEMP_HEAT_WARNING_3)
+		set_greyscale(colors=list(thermo_colors["orange"]))
 	else
-		icon_state_off = "heater"
-		icon_state_on = "heater_1"
-		icon_state_open = "heater-o"
+		set_greyscale(colors=list(thermo_colors["red"]))
+
```
Maybe it would be better to do this with a var and have only one case of set_greyscale at the end? might make things a bit cleaner

timothymtorres replied:
Agreed.  Fixed with ec061af06dd50d4e290e1f9ce46f5485f251beba

LemonInTheDark commented on code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm:
```diff
@@ -100,11 +118,16 @@
-		. += skipping
+		thermo_overlay.icon_state = "blinking"
+		. += new /mutable_appearance(thermo_overlay)
+		//var/mutable_appearance/skipping = mutable_appearance(thermo_overlay, "blinking")
+		//. += skipping
```
Did you intend to leave this in?

timothymtorres replied:
Missed this when I was cleaning up code.  Fixed with 5780de90a09af1ff23140910f23c2ae4155d7257

LemonInTheDark commented on code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm:
```diff
@@ -100,11 +118,16 @@
-	. += getpipeimage(icon, "pipe", turn(dir, 180), COLOR_MOSTLY_PURE_RED, piping_layer)
+	if(!initial(icon))
+		return
+	var/mutable_appearance/thermo_overlay = new(initial(icon))
```
Could this be cached?

timothymtorres replied:
I put the blinking overlay to use the GAGS JSON icon state instead with d3651b879263d290edebe1064f7523a9a2ad06dc

LemonInTheDark replied:
My question still applies, we're making an I think? identical mutable appearance for each thermomachine, could we cache it?

timothymtorres replied:
I missed your response.  The mutable appearance is for the pipe overlays which is something my PR doesn't touch.  I am not familiar with caching the overlays in DM.

If you are worried about performance, please consider that there are only ~5 thermomachines on a map so the speed increase would be _extremely meager_.  I have cached results for other code modules outside of DM & SS13, and when I benchmarked the results it's only worth it if there is a severe bottleneck in performance.  If there was something like thousands of instances, or a piece of code that was executed frequently then I could justify trying to boost the performance.

From my understanding of the mutable_apperance of the pipes below the thermomachine, it would have to cover all the pipe layers, directions, and colors.  I used the `!initial(icon)` code from the GAGS involving PDA's.  I'm still not even sure how it all works precisely.

LemonInTheDark replied:
if I'm honest it's just a gut reaction from seeing an overlay being created often. You're right, caching isn't a one size fits all tool, and this is more of a pattern thing. Maybe worth doing, but if you're uncomfortable with it we can skip for now

timothymtorres replied:
I'm going to opt to skip it.  Other than that, I believe everything else in this PR is ready to go.

LemonInTheDark commented on code/modules/atmospherics/machinery/other/meter.dm:
```diff
@@ -88,6 +102,24 @@
 	else
 		icon_state = "meter4"
 
+	var/env_temperature = environment.temperature
+	if(env_pressure == 0 || env_temperature == 0)
+		set_greyscale(colors=list(meter_colors["grey"]))
+	else if(env_temperature <= BODYTEMP_COLD_WARNING_3)
+		set_greyscale(colors=list(meter_colors["purple"]))
+	else if(env_temperature <= BODYTEMP_COLD_WARNING_2)
+		set_greyscale(colors=list(meter_colors["blue"]))
+	else if(env_temperature <= BODYTEMP_COLD_WARNING_1)
+		set_greyscale(colors=list(meter_colors["cyan"]))
+	else if(env_temperature <= BODYTEMP_HEAT_WARNING_1)
+		set_greyscale(colors=list(meter_colors["green"]))
+	else if(env_temperature <= BODYTEMP_HEAT_WARNING_2)
+		set_greyscale(colors=list(meter_colors["yellow"]))
+	else if(env_temperature <= BODYTEMP_HEAT_WARNING_3)
+		set_greyscale(colors=list(meter_colors["orange"]))
+	else
+		set_greyscale(colors=list(meter_colors["red"]))
```
Same thought about colors here

timothymtorres replied:
Agreed.  Fixed with ec061af06dd50d4e290e1f9ce46f5485f251beba

timothymtorres committed:
Remove commented out code overlay

timothymtorres committed:
Refactor greyscale colors to use variable instead of proc

Fix typo with greyscale var typo

timothymtorres committed:
Merge branch 'master' into atmos_mapping_upgrades

timothymtorres committed:
Add a cached thermomachine blinking icon state

timothymtorres committed:
Remove vars holding icon states

Since heater/freezer/thermomachine all uses the same sprite for GAGS there is no need to have these vars anymore.  Also changed the blinking to be a separate GAGS icon state.

timothymtorres committed:
Refactor greyscale colors for thermomachine and gas pipe meters

timothymtorres commented:
@LemonInTheDark I made the requested changes.  I also refactored the colors to use the #defines ones instead of creating new ones.

Twaticus reviewed (changes requested):
stmospherics/pipes/meter.dmi - buttons4 has a box on one of the frames
and for buttons0 its best to change the delay on the frame rather than copy paste the same frame multiple times in a row 
![image](https://user-images.githubusercontent.com/46540570/126050452-41e8c836-c4a6-40f9-83cf-8f25e1fa66df.png)

timothymtorres commented:
> stmospherics/pipes/meter.dmi - buttons4 has a box on one of the frames
> 
> and for buttons0 its best to change the delay on the frame rather than copy paste the same frame multiple times in a row 
> 
> ![image](https://user-images.githubusercontent.com/46540570/126050452-41e8c836-c4a6-40f9-83cf-8f25e1fa66df.png)
> 
> 

I was aware of this and both of these are intentional due to GAGS.  @ninjanomnom the GAGS icon animation doesn't allow a person to use a higher tick amount for a frame.  All the tick amounts revert to 1-tick rate.

Twaticus commented:
ah gotcha! i was unaware of the GAGs limitation

LemonInTheDark commented:
all good?

timothymtorres commented:
> all good?

If you are asking me, the PR is finished and ready.  👍

LemonInTheDark commented:
Let's jive
